sopel/modules/currency.py

```diff
@@ -2,109 +2,196 @@
-        rates = btc_rate.json()
-        return 1 / rates['averages']['day'], 'Bitcoin—24hr average'
 
-    data = get(base_url.format(code))
-    name = data.json()['seriesDetail']['FX{}CAD'.format(code)]['description']
-    name = name.split(" to Canadian")[0]
-    json = data.json()['observations']
-    for element in reversed(json):
-        if 'v' in element['FX{}CAD'.format(code)]:
-            return 1 / float(element['FX{}CAD'.format(code)]['v']), name
+class CurrencySection(StaticSection):
+    fixer_io_key = ValidatedAttribute('fixer_io_key', default=None)
+    auto_convert = ValidatedAttribute('auto_convert', parse=bool, default=False)
 
 
-@commands('cur', 'currency', 'exchange')
-@example('.cur 20 EUR in USD')
-def exchange(bot, trigger):
+def configure(config):
+    config.define_section('currency', CurrencySection, validate=False)
+    config.currency.configure_setting('fixer_io_key', 'Optional API key for Fixer.io (leave blank to use exchangeratesapi.io):')
+    config.currency.configure_setting('auto_convert', 'Automatically respond to currency conversion lines?')
+
+
+def setup(bot):
+    bot.config.define_section('currency', CurrencySection)
+
+
+def update_rates(bot):
+    global rates_fiat_json
+    global rates_btc_json
+
+    # If we have data that are less than 24h old, return
+    if 'date' in rates_fiat_json:
+        if time.time() - rates_fiat_json['date'] < 24 * 60 * 60:
+            return
+
+    # Update crypto rates
+    request = requests.get(CRYPTO_URL)
+    request.raise_for_status()
+    rates_btc_json = request.json()
+
+    # Update fiat rates
+    if bot.config.currency.fixer_io_key is not None:
+        request = requests.get(FIXER_URL.format(bot.config.currency.fixer_io_key))
+        if not request.json()['success']:
+            LOGGER.error(str(request.json()['error']))
+            return bot.reply('Sorry, something went wrong')
+    else:
+        request = requests.get(FIAT_URL)
+
+    request.raise_for_status()
+    rates_fiat_json = request.json()
+    rates_fiat_json['date'] = time.time()
+    rates_fiat_json['rates']['EUR'] = 1.0  # Put this here to make logic easier
+
+
+def btc_rate(code, reverse=False):
+    search = 'BTC{}'.format(code)
+
+    if search in rates_btc_json:
+        rate = rates_btc_json[search]['averages']['day']
+    else:
+        return "Sorry {} isn't currently supported".format(code)
+
+    if reverse:
+        return 1 / rate
+    else:
+        return rate
+
+
+def get_rate(of, to):
+    of = of.upper()
+    to = to.upper()
+
+    if of == 'BTC' or to == 'BTC':
+        if of == 'BTC':
+            code = to
+            reverse = False
+        else:
+            code = of
+            reverse = True
+
+        return btc_rate(code, reverse)
+
+    if of not in rates_fiat_json['rates']:
+        return "Sorry {} isn't currently supported".format(of)
+
+    if to not in rates_fiat_json['rates']:
+        return "Sorry {} isn't currently supported".format(to)
```

Comment:
Grammar nitpick:

```suggestion
        return "Sorry, {} isn't currently supported.".format(to)
```

(This is three times I made the same change to the same string… Might deserve to be a constant, e.g. `UNSUPPORTED_CURRENCY = "Sorry, {} isn't currently supported."`)